File: Problems/0001_0099/0038_Count_and_Say/Project_CS/Count_and_Say.cs
using System;

public class Solution
{
    public string CountAndSay(int n)
    {
        if ( n < 1 && n > 30)
        {
            return "";
        }

        string[] data = new string[n];

        data[0] = "1";
/*      data[1] = "11";
        data[2] = "21";
        data[3] = "1211";
        data[4] = "111221";
        data[5] = "312211";
        data[6] = "13112221";
        data[7] = "1131213211";
*/

        int i, pos, count;

        for ( i = 1; i < n; i++ )
        {
            data[i] = "";
            for ( pos = 0; pos < data[i - 1].Length; pos += count )
            {
                count = count_continuity_num(data[i - 1], pos);
                data[i] += count.ToString() + data[i - 1][pos];
            }
        }
        
        return ( data[n - 1] );
    }
    
    private int count_continuity_num(string data, int pos)
    {
        int i = 0, count = 0;
        
        for ( ; (pos + i) < data.Length; ++i ) {
            if ( data[pos + i] == data[pos] ) {
                count++;
            }
            else {
                return count;
            }
        }
        
        return count;
    }

    public void Main(string args)
    {
        string fld = args.Replace("[","").Replace("]","").Trim();
        int n = int.Parse(fld);

        Console.WriteLine("n = " + n);

        System.Diagnostics.Stopwatch sw = new System.Diagnostics.Stopwatch();
        sw.Start();

        string result = CountAndSay(n);
        Console.WriteLine("result = " + result);
        
        sw.Stop();
        Console.WriteLine("Execute time ... " + sw.ElapsedMilliseconds.ToString() + "ms\n");
    }
}

File: Problems/0001_0099/0038_Count_and_Say/Project_Python3/Count_and_Say.py
import os
import sys
import time
from typing import List, Dict, Tuple

class Solution:
    def countAndSay(self, n: int) -> str:
        # 41ms - 66ms
        if n == 1:
            return "1"
        x = self.countAndSay(n - 1)
        ans, y, ct = "", x[0], 1
        for i in range(1, len(x)):
            if x[i] == y:
                ct += 1
            else:
                ans += str(ct) + str(y)
                y, ct = x[i], 1
        ans += str(ct) + str(y)
        return ans

    def countAndSay2(self, n: int) -> str:
        # 76ms - 83ms
        if n < 1 and n > 30:
            return ""
        data = []
        data.append("1")
        for _ in range(1, n):
            temp = ""
            pos = 0
            while pos < len(data[-1]):
                count = self.count_continuity_num(data[-1], pos)
                temp += str(count) + data[-1][pos]
                pos += count
            data.append(temp)
        return data[n - 1]

    def count_continuity_num(self, data: str, pos: int) -> int:
        i, count = 0, 0
        while pos + i < len(data):
            if data[pos + i] == data[pos]:
                count += 1
            else:
                return count
            i += 1
        return count
 
def main():
    argv = sys.argv
    argc = len(argv)

    if argc < 2:
        print("Usage: python {0} <testdata.txt>".format(argv[0]))
        exit(0)

    if not os.path.exists(argv[1]):
        print("{0} not found...".format(argv[1]))
        exit(0)

    testDataFile = open(argv[1], "r")
    lines = testDataFile.readlines()

    for temp in lines:
        temp = temp.strip()
        if temp == "":
            continue
        print("args = {0}".format(temp))
        loop_main(temp)
    #   print("Hit Return to continue...")
    #   input()

def loop_main(temp):
    fld = temp.replace("\"","").replace("[","").replace("]","").rstrip()
    n = int(fld)
    print("n = {0:d}".format(n))

    sl = Solution()
    time0 = time.time()

    result = sl.countAndSay(n)

    time1 = time.time()

    print("result = \"{0}\"".format(result))
    print("Execute time ... : {0:f}[s]\n".format(time1 - time0))

if __name__ == "__main__":
    main()
